source/pervasive/arithmetic/mul.rs

```diff
@@ -0,0 +1,757 @@
+pub proof fn lemma_mul_nonzero(x: int, y: int)
+    ensures
+        x * y != 0 <==> x != 0 && y != 0,
+{
+    MulINL::lemma_mul_nonzero(x, y);
+}
+
+/// Proof that a product is nonzero if and only if both factors are nonzero
+pub proof fn lemma_mul_nonzero_auto()
+    ensures
+        forall|x: int, y: int| #[trigger] (x * y) != 0 <==> x != 0 && y != 0,
+{
+    assert forall|x: int, y: int| #[trigger] (x * y) != 0 <==> x != 0 && y != 0 by {
+        lemma_mul_nonzero(x, y);
+    }
+}
+
+/// Proof that any integer multiplied by 0 results in a product of 0
+pub proof fn lemma_mul_by_zero_is_zero_auto()
+    ensures
+        forall|x: int| #![trigger x * 0] #![trigger 0 * x] x * 0 == 0 && 0 * x == 0,
+{
+    assert forall|x: int| #![trigger x * 0] #![trigger 0 * x] x * 0 == 0 && 0 * x == 0 by {
+        lemma_mul_basics(x);
+    }
+}
+
+/// Proof that multiplication is associative, specifically that
+/// `x * (y * z) == (x * y) * z`.
+pub proof fn lemma_mul_is_associative(x: int, y: int, z: int)
+    ensures
+        x * (y * z) == (x * y) * z,
+{
+    MulINL::lemma_mul_is_associative(x, y, z);
+}
+
+/// Proof that multiplication is associative
+pub proof fn lemma_mul_is_associative_auto()
+    ensures
+        forall|x: int, y: int, z: int|
+            #![trigger x * (y * z)]
+            #![trigger (x * y) * z]
+            (x * (y * z)) == ((x * y) * z),
+{
+    assert forall|x: int, y: int, z: int|
+        #![trigger x * (y * z)]
+        #![trigger (x * y) * z]
+        (x * (y * z)) == ((x * y) * z) by {
+        lemma_mul_is_associative(x, y, z);
+    }
+}
+
+/// Proof that multiplication is commutative, specifically that
+/// `x * y == y * x`.
+pub proof fn lemma_mul_is_commutative(x: int, y: int)
+    ensures
+        x * y == y * x,
+{
+}
+
+/// Proof that multiplication is commutative
+pub proof fn lemma_mul_is_commutative_auto()
+    ensures
+        forall|x: int, y: int| #[trigger] (x * y) == (y * x),
+{
+}
+
+/// Proof that, since the product of the two integers `x` and `y` is
+/// nonnegative, that product is greater than or equal to each of `x`
+/// and `y`
+pub proof fn lemma_mul_ordering(x: int, y: int)
+    requires
+        x != 0,
+        y != 0,
+        0 <= x * y,
+    ensures
+        x * y >= x && x * y >= y,
+{
+    MulINL::lemma_mul_ordering(x, y);
+}
+
+/// Proof that if the product of two integers is nonnegative, then
+/// that product is greater than or equal to each of the two integers
+proof fn lemma_mul_ordering_auto()
+    ensures
+        forall|x: int, y: int|
+            (0 != x && 0 != y && #[trigger]
+            (x * y) >= 0) ==> x * y >= x && (x * y) >= y,
+{
+    assert forall|x: int, y: int| 0 != x && 0 != y && x * y >= 0 implies #[trigger]
+    (x * y) >= x && #[trigger]
+    (x * y) >= y by {
+        lemma_mul_ordering(x, y);
+    };
+}
+
+/*
+    We don't port LemmaMulEquality or LemmaMulEqualityAuto from the
+    Dafny standard library for arithmetic, since they're never useful.
```

Comment:
This is a bit enigmatic.  Why are they never useful?

Reply:
I fleshed out the comment more to clarify, and got Chris's help with the wording.